lib/tools/apks-utils.js

```diff
@@ -0,0 +1,210 @@
+      log.debug(`Command stderr: ${e.stderr}`);
+    }
+    throw new Error(`${errorMsg}. Original error: ${e.message}`);
+  }
+};
+
+/**
+ * @param {string} specLocation - The full path to the generated device spec location
+ * @returns {string} The same `specLocation` value
+ * @throws {Error} If it is not possible to retrieve the spec for the current device
+ */
+apksUtilsMethods.getDeviceSpec = async function (specLocation) {
+  const args = [
+    'get-device-spec',
+    '--adb', this.executable.path,
+    '--device-id', this.curDeviceId,
+    '--output', specLocation,
+  ];
+  log.debug(`Getting the spec for the device '${this.curDeviceId}'`);
+  await this.execBundletool(args, 'Cannot retrieve the device spec');
+  return specLocation;
+};
+
+/**
+ * @typedef {Object} InstallApksOptions
+ * @property {?number|string} timeout [20000] - The number of milliseconds to wait until
+ *                                              the installation is completed
+ * @property {boolean} allowTestPackages [false] - Set to true in order to allow test
+ *                                                 packages installation.
+ * @property {boolean} useSdcard [false] - Set to true to install the app on sdcard
+ *                                         instead of the device memory.
+ * @property {boolean} grantPermissions [false] - Set to true in order to grant all the
+ *                                                permissions requested in the application's manifest
+ *                                                automatically after the installation is completed
+ *                                                under Android 6+.
+ */
+
+/**
+ * Installs the given .apks package into the device under test
+ *
+ * @param {string} apks - The full path to the .apks file
+ * @param {?InstallApksOptions} options - Installation options
+ * @throws {Error} If the .apks bundle cannot be installed
+ */
+apksUtilsMethods.installApks = async function (apks, options = {}) {
+  options = Object.assign({
+    timeout: APKS_INSTALL_TIMEOUT,
+  }, options, {replace: true});
+
+  const tmpRoot = await tempDir.openDir();
+  try {
+    const specPath = await this.getDeviceSpec(path.resolve(tmpRoot, 'deviceSpec.json'));
+    const args = [
+      'extract-apks',
+      '--adb', this.executable.path,
+      '--apks', apks,
+      '--output-dir', tmpRoot,
+      '--device-spec', specPath,
+    ];
+    log.debug(`Extracting the apk files from '${apks}'`);
+    await this.execBundletool(args, `Cannot extract the application bundle at '${apks}'`);
+    const installArgs = buildInstallArgs(await this.getApiLevel(), options);
+    const apksToInstall = (await fs.readdir(tmpRoot))
+      .filter((name) => name.endsWith('.apk'))
+      .map((name) => path.resolve(tmpRoot, name));
+    log.debug(`Got the following apk files to install: ${JSON.stringify(apksToInstall)}`);
+    const output = await this.adbExec(['install-multiple', ...installArgs, ...apksToInstall], {
```

Comment:
If the install apks has unsuitable architecture, installing apks fail. (`adb install` failed before. It happen if users use ML related libraries, for example.)

```
base-vi.apk 11K
base-x86.apk 483K
base-x86_64.apk 453K
base-xhdpi.apk 2.4M
```

~~What about add considering abis?~~
=> ah, we use `'--device-spec', specPath,`. nice 👍